subsys/net/ip/tcp.c

```diff
@@ -415,7 +415,20 @@ static int tcp_conn_unref(struct tcp *conn, int status)
 	if (conn->in_connect) {
 		NET_DBG("conn: %p is waiting on connect semaphore", conn);
 		tcp_send_queue_flush(conn);
-		goto out;
+
+		if (conn->async_connect) {
```

Comment:
I see the problem of feeding  `connect_sem` in a sychronized case when it would wake up the thread waiting for `connect_sem` in `net_tcp_connect` and then it would race in the `tcp_conn_unref` . Though in that case, there may be a solution by moving `tcp_lock` to the top of that function. What do you think?

Reply:
It seems that the code around the `in_connect` flag could use some extra mutex protection anyway. It seems to me that a similar race could take place when you close the socket that connects asynchronously before the connection is established.

Reply:
Ok I used `conn->lock` to protect that section